jsign-crypto/src/main/java/net/jsign/jca/VenafiSigningService.java

```diff
@@ -0,0 +1,269 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.jsign.jca;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.KeyStoreException;
+import java.security.UnrecoverableKeyException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.LinkedHashMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.io.ByteArrayOutputStream;
+import java.math.BigInteger;
+import org.bouncycastle.asn1.ASN1Encodable;
+import org.bouncycastle.asn1.DLSequence;
+import org.bouncycastle.asn1.ASN1Integer;
+import org.bouncycastle.asn1.ASN1OutputStream;
+
+import net.jsign.DigestAlgorithm;
+
+/**
+ * Signing service using the Venafi CodeSign Protect REST Service API.
+ *
+ * @since 7.2
+ * @see <a href="https://docs.venafi.com/Docs/current/TopNav/Content/SDK/CodeSignSDK/cco-SDKc-Code-Signing-SDK-reference.php">Venafi CodeSign Protect REST API</a>
+ */
+public class VenafiSigningService implements SigningService {
+
+    /** Cache of certificates indexed by alias */
+    private final Map<String, Map<String, ?>> certificates = new LinkedHashMap<>();
+
+    /** The API endpoint of the Venafi CodeSign Protect */
+    private final String endpoint;
+
+    private final RESTClient client;
+
+    /** The credentials to authenticate with the service */
+    private final VenafiCredentials credentials;
+
+     /** The name of the Venafi CodeSign Protect certificate KeyId */
+     private String KeyId;
+
+    /** Mapping between Java and Venafi CodeSign Protect signing algorithms */
+    private final Map<String, Integer> algorithmMapping = new HashMap<>();
+    {
+        algorithmMapping.put("SHA256withRSA", 64);
+        algorithmMapping.put("SHA384withRSA", 65);
+        algorithmMapping.put("SHA512withRSA", 66);
+        algorithmMapping.put("SHA256withECDSA", 4164);
+        algorithmMapping.put("SHA384withECDSA", 4165);
+        algorithmMapping.put("SHA512withECDSA", 4166);
+    }
+
+    /* Map ASN.1 DER prefix structures to MessageDigest Algorithm */
+
+    private final byte[] getHashPrefix(Integer mechanism) {
+        switch (mechanism) {
+            case 64: case 4164:
+                return new byte[]{0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, (byte)0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20}; // SHA256
```

Comment:
Couldn't we use BouncyCastle to generate the ASN.1 structure instead of assembling the bytes manually?